### ISSUE TITLE
Errors reported by ConstantEvaluationValidator should be made more specific

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#47061]
Right now most errors use the default error, "Const variables must be initialized with a constant value." (const_initialized_with_non_constant_value). Each case can be given more information though, .e.g "constructor reference used in constant expression must be a const constructor."

**Comment by bwilkerson** [comment]
Yes, and many of them should have context messages associated with them. I took a look at this a couple of years ago but didn't have time for the amount of work that I thought would be necessary. Happy to discuss if you're interested.

**Comment by asashour** [comment]
The only usage I see is

https://github.com/dart-lang/sdk/blob/a6cc772b670034f355c626e28ce11d1555f97e24/pkg/analyzer/lib/src/dart/constant/constant_verifier.dart#L249-L250

This is a new area for me, but am interested to help if further requirements are specified.

**Comment by srawlins** [comment]
@kallentu tackled this and improved a lot of error categories. Closing this in favor of any remaining specific issues.
